packages/execution-environments/src/common/BaseSnapExecutor.ts

```diff
@@ -0,0 +1,250 @@
+      this.respond(id, {
+        error: ethErrors.rpc
+          .methodNotFound({
+            data: {
+              method,
+            },
+          })
+          .serialize(),
+      });
+      return;
+    }
+
+    // support params by-name and by-position
+    const paramsAsArray = sortParamKeys(methodObject, params);
+
+    try {
+      const result = await (this.methods as any)[method](...paramsAsArray);
+      this.respond(id, { result });
+    } catch (e) {
+      this.respond(id, {
+        error: serializeError(e, {
+          fallbackError,
+        }),
+      });
+    }
+  }
+
+  protected notify(requestObject: Record<string, unknown>) {
+    this.commandStream.write({
+      ...requestObject,
+      jsonrpc: '2.0',
+    });
+  }
+
+  protected respond(id: JSONRPCID, responseObj: Record<string, unknown>) {
+    this.commandStream.write({
+      ...responseObj,
+      id,
+      jsonrpc: '2.0',
+    });
+  }
+
+  /**
+   * Attempts to evaluate a snap in SES.
+   * Generates the APIs for the snap. May throw on error.
+   *
+   * @param {string} snapName - The name of the snap.
+   * @param {Array<string>} approvedPermissions - The snap's approved permissions.
+   * Should always be a value returned from the permissions controller.
+   * @param {string} sourceCode - The source code of the snap, in IIFE format.
+   * @param {Array} endowments - An array of the names of the endowments.
+   */
+  protected startSnap(
+    snapName: string,
+    sourceCode: string,
+    _endowments?: Endowments,
+  ) {
+    console.log(`starting snap '${snapName}' in worker`);
+    if (this.snapPromiseErrorHandler) {
+      self?.removeEventListener(
+        'unhandledrejection',
+        this.snapPromiseErrorHandler,
+      );
+    }
+
+    if (this.snapErrorHandler) {
+      self?.removeEventListener('error', this.snapErrorHandler);
+    }
+
+    const wallet = this.createSnapProvider(snapName);
+
+    const endowments: Record<string, any> = {
+      BigInt,
+      Buffer,
+      console,
+      crypto: self.crypto,
+      Date,
+      Math,
+      SubtleCrypto: self.SubtleCrypto,
+      wallet,
+      ...createTimeout(),
+    };
+
+    (_endowments || []).forEach((_endowment) => {
+      if (!(_endowment in self)) {
+        throw new Error(`Unknown endowment: "${_endowment}".`);
+      }
+
+      const globalValue = (self as any)[_endowment];
+      endowments[_endowment] =
+        typeof globalValue === 'function'
+          ? globalValue.bind(self)
+          : globalValue;
+    });
+
+    this.snapErrorHandler = (error: ErrorEvent) => {
+      this.errorHandler(error.error, { snapName });
+    };
+
+    this.snapPromiseErrorHandler = (error: PromiseRejectionEvent) => {
+      this.errorHandler(error.reason, { snapName });
+    };
+
+    try {
+      const compartment = new Compartment({
+        ...endowments,
+        window: { ...endowments },
+      });
+      compartment.evaluate(sourceCode);
+
+      self?.addEventListener(
```

Comment:
follow-up: We should fix the types such that `self` is always defined, if possible.